### PR TITLE
Fix inconsistent of type alias 

### LINKED ISSUE #1507: Inconsistent use of type alias.
Inside `basic_json` the alias `reference` and `const_reference` are used in inconsistently, for example the method `iteration_proxy` uses `const_reference` as a parameter type, while `push_back` uses `const basic_json&`.
Which one should be used? I think the non-alias are better and i would like make a PR if you agree.

Sorry about my horrible english.

### PR DESCRIPTION
**Pull request by DodoMomo**
Fixes #1507

### CODE REVIEW

**Comment by nlohmann** [comment]
Sorry, I messed up the current develop branch. I shall commit a fix in a second.

**Comment by nlohmann** [comment]
Please rebase to 37a72dac48035e12df13c923e19f10cbe8a22f5f.

**Comment by nlohmann** [comment]
FYI:
https://github.com/nlohmann/json/issues/1507#issuecomment-473994405

File: include/nlohmann/json.hpp [hunk 1]
```diff
@@ -2912,7 +2912,7 @@ class basic_json
     written using `at()`. It also demonstrates the different exceptions that
     can be thrown.,at__size_type}
     */
-    reference at(size_type idx)
+    basic_json& at(size_type idx)
```
Comment (nlohmann) [inline include/nlohmann/json.hpp]: I think the change from `reference` to `basic_json&` is not improving readability. Unless there are good reasons, I would not approve this PR.

**Comment by DodoMomo** [comment]
So what do you think is the right procedure for type alias? Only parts of the STL-like interface? Every place where the it refers to the element type?

**Comment by nlohmann** [comment]
I would say each occurrence of (const) `basic_json&` in the interface should be replaced by (const) `reference`.

**Pull request closed without merging**